perception/sub8_perception/src/sub8_perception/start_gate.cpp

```diff
@@ -10,66 +10,60 @@ Sub8StartGateDetector::~Sub8StartGateDetector() {}
-
-            //         //Make bigger buffer. After like 30 calculate mean and average. Outlier stuff.
-            //     }
-            // }
-            for (size_t i = 0; i < gate_line_buffer[29].size(); i++) {
-                cv::Point2i s = gate_line_buffer[29][i][1] - gate_line_buffer[29][i][0];
+
+            //Look at one of the elements (gate_line_buffer.capacity()-1) in the circular buffer and check it's statistics and delete if necessary
+            for (size_t i = 0; i < gate_line_buffer[gate_line_buffer.capacity()-1].size(); i++) {
+                cv::Point2i s = gate_line_buffer[gate_line_buffer.capacity()-1][i][1] - gate_line_buffer[gate_line_buffer.capacity()-1][i][0];
                 cv::Point2i center = cv::Point2i(s.x / 2 + s.x, s.y / 2 + s.y);
-                // std::cout << center << std::endl;
+
+                //If position of the element is greater than one deviation from the mean, then delete
                 if (abs(center.x - mean(accX)) > sqrt(variance(accX)) ||
                         abs(center.y - mean(accY) > sqrt(variance(accY)))) {
-                    gate_line_buffer[29].erase(gate_line_buffer[29].begin() + i);
+                    gate_line_buffer[gate_line_buffer.capacity()-1].erase(gate_line_buffer[gate_line_buffer.capacity()-1].begin() + i);
                     i--;
                     std::cout << "Deleted: " << center << std::endl;
                     continue;
                 }
+
+                //If the size of the element is greater than one deviation from the mean, then delete
                 if (abs(s.x - mean(accSizeX)) > sqrt(variance(accSizeX)) ||
                         abs(s.y - mean(accSizeY)) > sqrt(variance(accSizeY))) {
-                    gate_line_buffer[29].erase(gate_line_buffer[29].begin() + i);
+                    gate_line_buffer[gate_line_buffer.capacity()-1].erase(gate_line_buffer[gate_line_buffer.capacity()-1].begin() + i);
                     i--;
                     std::cout << "Deleted size: " << s << std::endl;
                     continue;
                 }
             }
-            cv::Mat show = cv::Mat::zeros(rows, cols, CV_8UC3);
 
-            for (size_t i = 0; i < gate_line_buffer[29].size(); i++) {
-                // for (int j = 0; j < gate_line_buffer[0][i].size(); j++) {
-                // cv::line(show, gate_line_buffer[0][i][j], gate_line_buffer[0][i][(j + 1) % 4], cv::Scalar(0, 0, 255), 1, 8 );
-                // }
-                cv::rectangle(show, gate_line_buffer[29][i][0], gate_line_buffer[29][i][1], cv::Scalar(255, 0, 0), 2, 8, 0);
+            //Visualize on of the saved gates in the buffer
+            
+// #ifdef VISUALIZE
+            cv::Mat show = cv_bridge::toCvShare(image_msg, "bgr8")->image;
+            for(auto &gate : gate_line_buffer[29]) {
+                cv::rectangle(show, gate[0], gate[1], cv::Scalar(255, 0, 0), 2, 8, 0);
             }
             cv::imshow("ay", show);
```

Comment:
Instead of using #ifdef visualize and opening gui windows, it would be cleaner to publish visualization images.